spec/pg/decoders/range_decoder_spec.cr

```
@@ -0,0 +1,109 @@
+      test_decode "empty range", "'empty'::int8range", 0_i64...0_i64
+      test_decode "(-infinity,upper] - infinite lower bound", "'(,4000000000]'::int8range", nil...4_000_000_001
+      test_decode "[lower,+infinity) - infinite upper bound", "'[3000000000,)'::int8range", 3_000_000_000...nil
+      test_decode "(-infinity,+infinity) - both bounds infinite", "'(,)'::int8range", nil...nil
+    end
+
+    describe "daterange" do
+      test_decode "(lower,upper) - both exclusive", "'(2023-01-01,2023-12-31)'::daterange", Time.utc(2023, 1, 2)...Time.utc(2023, 12, 31)
+      test_decode "(lower,upper] - exclusive lower, inclusive upper", "'(2023-01-01,2023-12-31]'::daterange", Time.utc(2023, 1, 2)...Time.utc(2024, 1, 1)
+      test_decode "[lower,upper) - inclusive lower, exclusive upper", "'[2023-01-01,2023-12-31)'::daterange", Time.utc(2023, 1, 1)...Time.utc(2023, 12, 31)
+      test_decode "[lower,upper] - both inclusive", "'[2023-01-01,2023-12-31]'::daterange", Time.utc(2023, 1, 1)...Time.utc(2024, 1, 1)
+      test_decode "empty range", "'empty'::daterange", Time.unix(0)...Time.unix(0)
+      test_decode "(-infinity,upper] - infinite lower bound", "'(,2023-12-31]'::daterange", nil...Time.utc(2024, 1, 1)
+      test_decode "[lower,+infinity) - infinite upper bound", "'[2023-01-01,)'::daterange", Time.utc(2023, 1, 1)...nil
+      test_decode "(-infinity,+infinity) - both bounds infinite", "'(,)'::daterange", nil...nil
+    end
+
+    describe "tsrange" do
+      test_decode "(lower,upper) - both exclusive", "'(2023-01-01 10:30:00,2023-12-31 15:45:00)'::tsrange", Time.utc(2023, 1, 1, 10, 30, 0)...Time.utc(2023, 12, 31, 15, 45, 0)
+      test_decode "(lower,upper] - exclusive lower, inclusive upper", "'(2023-01-01 10:30:00,2023-12-31 15:45:00]'::tsrange", Time.utc(2023, 1, 1, 10, 30, 0)..Time.utc(2023, 12, 31, 15, 45, 0) # Crystal can't represent exclusive lower
```

**issue:** Changing the exclusive lower to an inclusive one is a dangerous change.
Unfortunately, we cannot represent that with `Range`. But I'd consider silently converting the range to different semantics an error. 

For `Time` we can work around by treating it as a discrete type. Just add one nanosecond:
```suggestion
      test_decode "(lower,upper] - exclusive lower, inclusive upper", "'(2023-01-01 10:30:00,2023-12-31 15:45:00]'::tsrange", Time.utc(2023, 1, 1, 10, 30, 0, nanosecond: 1)..Time.utc(2023, 12, 31, 15, 45, 0)
```
This doesn't preserve the _exact_ semantics either, but it's much closer and preserves the _intent_.

Alternatively, we need a different solution (e.g. a custom `Range` type).
Omitting this case (i.e. raise an error) would also be better than implicitly changing semantics.

---

Right. Silently converting exclusive lower bounds to inclusive ones changes the semantics and could lead to bugs.

https://www.postgresql.org/docs/current/rangetypes.html#RANGETYPES-DISCRETE

- Discrete types (`int4`, `int8`, `date`): have clear "next" values (`1 -> 2`,  `2023-01-01` -> `2023-01-02`). This is working.
- Continuous types (`timestamp`, `numeric`): don't have clear "next" values.

From the documentation:

> Even though timestamp has limited precision, and so could theoretically be treated as discrete, it's better to consider it continuous since the step size is normally not of interest.

Do you think I should implement these adjustments?